pkg/cloud/openstack/services/certificates/certificates.go

```diff
@@ -0,0 +1,334 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog"
+	"math"
+	"math/big"
+	"net"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+	"strings"
+	"time"
+)
+
+const (
+	rsaKeySize     = 2048
+	duration365d   = time.Hour * 24 * 365
+	clusterCA      = "cluster-ca"
+	etcdCA         = "etcd-ca"
+	frontProxyCA   = "front-proxy-ca"
+	serviceAccount = "service-account"
+)
+
+// NewPrivateKey creates an RSA private key
+func NewPrivateKey() (*rsa.PrivateKey, error) {
+	return rsa.GenerateKey(rand.Reader, rsaKeySize)
+}
+
+// AltNames contains the domain names and IP addresses that will be added
+// to the API Server's x509 certificate SubAltNames field. The values will
+// be passed directly to the x509.Certificate object.
+type AltNames struct {
+	DNSNames []string
+	IPs      []net.IP
+}
+
+// Config contains the basic fields required for creating a certificate
+type Config struct {
+	CommonName   string
+	Organization []string
+	AltNames     AltNames
+	Usages       []x509.ExtKeyUsage
+}
+
+// ReconcileCertificates generate certificates if none exists.
+func (s *Service) ReconcileCertificates(clusterName string, clusterProviderSpec *v1alpha1.OpenstackClusterProviderSpec) error {
+	if !clusterProviderSpec.CAKeyPair.HasCertAndKey() {
```

Comment:
from code, looks like you are trying to generate those keys? if it's generated, then how do you store them and reuse? looks to me the Spec are the desired state but if you didn't give key at beginning, it's not desired state?? just curious ..

Reply:
I'll move this PR and your comment over to the kubernetes-sigs repo now that the first PR is merged :)

Reply:
Moved here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/394/files#r302672418